File: src/core/tools/select.ts
// Entity-aware select & move tool.
// Ported from ASCIIFlow (client/draw/select.ts), MIT © Lewis Hemens, plus
// lazydraw's keyboard nudge, clipboard and paste (§5.2).
import { Box, boundingBox } from "../box";
import { isBoxDrawing } from "../glyphs";
import {
  type BoxAttachment,
  cellsInBox,
  detectLineTip,
  detectWord,
  findBox,
  moveBoxWithAttachments,
  moveCells,
  traceBoxAttachments,
  traceLineFromTip,
} from "../entity";
import { Layer, StackedLayers } from "../layer";
import { arrowFor, connectEndpoints, line } from "../route";
import { snap } from "../snap";
import { layerToText, textToLayer } from "../text";
import { Vector } from "../vector";
import { MoveTool } from "./move";
import { KEY, type HoverHint, type Mods, type Tool, type ToolContext } from "./tool";

interface LineReshape {
  cells: Vector[];
  anchor: Vector;
  isArrow: boolean;
  horizontalSegment: boolean;
  moved: boolean;
}

const NUDGE: Record<string, Vector> = {
  [KEY.UP]: new Vector(0, -1),
  [KEY.DOWN]: new Vector(0, 1),
  [KEY.LEFT]: new Vector(-1, 0),
  [KEY.RIGHT]: new Vector(1, 0),
};

function dedupe(cells: Vector[]): Vector[] {
  const seen = new Set<string>();
  return cells.filter((c) => !seen.has(c.key()) && (seen.add(c.key()), true));
}

export class SelectTool implements Tool {
  selectBox: Box | null = null;
  private selectedCells: Vector[] = [];
  private selecting = false;
  private selectAnchor: Vector | null = null;
  private dragStart: Vector | null = null;
  private dragEnd: Vector | null = null;
  private moveTool: MoveTool | null = null;
  private lineReshape: LineReshape | null = null;
  private activeBox: Box | null = null;
  private attachments: BoxAttachment[] = [];

  constructor(private readonly ctx: ToolContext) {}

  private get canvas() {
    return this.ctx.canvas;
  }

  get hasSelection(): boolean {
    return this.selectedCells.length > 0 && this.canvas.selection != null;
  }

  start(p: Vector): void {
    const committed = this.canvas.committed;
    const value = committed.get(p);

    if (this.selectedCells.length > 0 && this.inSelection(p)) {
      this.beginDrag(p);
      return;
    }

    const tip = detectLineTip(committed, p);
    if (tip) {
      const { cells, anchor } = traceLineFromTip(committed, tip.tip, tip.bodyDir);
      this.lineReshape = {
        cells,
        anchor,
        isArrow: tip.arrow != null,
        horizontalSegment: tip.axis === "horizontal",
        moved: false,
      };
      return;
    }

    const word = detectWord(committed, p);
    if (word) {
      this.setSelection(word);
      this.beginDrag(p);
      return;
    }

    if (isBoxDrawing(value)) {
      this.moveTool = new MoveTool(this.ctx);
      this.moveTool.start(p);
      return;
    }

    const box = findBox(committed, p);
    if (box) {
      this.setSelection(cellsInBox(committed, box));
      this.activeBox = box;
      this.beginDrag(p);
      return;
    }

    this.startSelect(p);
  }

  move(p: Vector, m: Mods): void {
    if (this.lineReshape) this.reshapeTip(p, m);
    else if (this.dragStart) this.moveDrag(p);
    else if (this.moveTool) this.moveTool.move(p);
    else if (this.selecting) this.moveSelect(p);
  }

  end(): void {
    if (this.lineReshape) {
      if (this.lineReshape.moved) this.canvas.commitScratch();
      else this.canvas.clearScratch();
      this.lineReshape = null;
    } else if (this.dragStart && this.dragEnd) {
      const delta = this.dragEnd.subtract(this.dragStart);
      const box = this.activeBox;
      this.canvas.commitScratch();
      if (box) {
        const moved = box.translate(delta);
        this.selectedCells = cellsInBox(this.canvas.committed, moved);
        this.selectBox = moved;
        this.activeBox = moved;
        this.canvas.setSelection(moved);
      } else {
        this.setSelection(
          this.selectedCells.map((c) => c.add(delta)),
          true,
        );
      }
    } else if (this.moveTool) {
      this.moveTool.end();
      this.moveTool = null;
    } else if (this.selecting) {
      this.finishSelect();
    }
    this.dragStart = null;
    this.dragEnd = null;
    this.selecting = false;
  }

  private setSelection(cells: Vector[], keepScratch = false): void {
    this.selectedCells = dedupe(cells);
    this.selectBox = boundingBox(this.selectedCells);
    this.activeBox = null;
    if (!keepScratch) this.canvas.clearScratch();
    this.canvas.setSelection(this.selectBox);
  }

  private inSelection(p: Vector): boolean {
    return this.selectBox != null && this.canvas.selection != null && this.selectBox.contains(p);
  }

  private startSelect(p: Vector): void {
    this.selecting = true;
    this.selectAnchor = p;
    this.selectBox = new Box(p, p);
    this.selectedCells = [];
    this.activeBox = null;
    this.canvas.setSelection(this.selectBox);
  }

  private moveSelect(p: Vector): void {
    this.selectBox = new Box(this.selectAnchor!, p);
    this.canvas.setSelection(this.selectBox);
  }

  private finishSelect(): void {
    this.selectedCells = cellsInBox(this.canvas.committed, this.selectBox!);
    // A rubber-band selection moves like a box: lines crossing its edge reflow.
    this.activeBox = this.selectBox;
  }

  private beginDrag(p: Vector): void {
    this.dragStart = p;
    this.dragEnd = p;
    this.attachments = this.activeBox ? traceBoxAttachments(this.canvas.committed, this.activeBox) : [];
  }

  private moveDrag(p: Vector): void {
    this.dragEnd = p;
    const delta = p.subtract(this.dragStart!);
    const committed = this.canvas.committed;
    if (this.activeBox) {
      this.canvas.setScratch(moveBoxWithAttachments(committed, this.activeBox, this.attachments, delta));
      this.canvas.setSelection(this.activeBox.translate(delta));
    } else {
      this.canvas.setScratch(moveCells(committed, this.selectedCells, delta));
      this.canvas.setSelection(boundingBox(this.selectedCells.map((c) => c.add(delta))));
    }
  }

  private reshapeTip(target: Vector, m: Mods): void {
    const r = this.lineReshape!;
    r.moved = true;
    const base = this.canvas.committed.clone();
    for (const c of r.cells) base.delete(c);

    const layer = new Layer();
    for (const c of r.cells) layer.set(c, "");

    if (!r.anchor.equals(target)) {
      // Leave the pivot along the line's axis, then bend toward the cursor.
      const horizontalFirst = r.horizontalSegment !== m.flip;
      layer.setFrom(line(r.anchor, target, horizontalFirst));
      if (r.isArrow) layer.set(target, arrowFor(r.anchor, target, horizontalFirst));
      connectEndpoints(layer, base, r.isArrow ? [r.anchor] : [r.anchor, target]);
    }

    layer.setFrom(snap(layer, base));
    this.canvas.setScratch(layer);
    this.canvas.clearSelection();
  }

  hoverHint(p: Vector): HoverHint {
    const committed = this.canvas.committed;
    if (this.selectedCells.length > 0 && this.inSelection(p)) return "move";
    if (detectLineTip(committed, p)) return "crosshair";
    if (detectWord(committed, p)) return "move";
    if (isBoxDrawing(committed.get(p))) return new MoveTool(this.ctx).hoverHint(p);
    if (findBox(committed, p)) return "move";
    return "default";
  }

  /** Text of the current selection, or null. */
  copySelection(): string | null {
    if (!this.selectBox || !this.canvas.selection) return null;
    const cells = new Layer();
    for (const c of this.selectedCells) {
      const v = this.canvas.committed.get(c);
      if (v) cells.set(c, v);
    }
    return layerToText(new StackedLayers([cells]), this.selectBox);
  }

  /** Erases the selected content as one undo step. */
  eraseSelection(): boolean {
    if (this.selectedCells.length === 0) return false;
    const layer = new Layer();
    for (const c of this.selectedCells) layer.set(c, "");
    layer.setFrom(snap(layer, this.canvas.committed));
    const changed = this.canvas.commit(layer);
    this.selectedCells = [];
    this.selectBox = null;
    this.activeBox = null;
    this.canvas.clearSelection();
    return changed;
  }

  /** Pastes text with its top-left at `at`, committed, and selects it. */
  paste(text: string, at: Vector): void {
    const pasted = textToLayer(text, at);
    if (pasted.size() === 0) return;
    const layer = pasted.clone();
    layer.setFrom(snap(pasted, this.canvas.committed));
    this.canvas.commit(layer);
    this.setSelection(pasted.keys());
  }

  handleKey(key: string): boolean {
    if (key === KEY.BACKSPACE || key === KEY.DELETE) {
      return this.eraseSelection();
    }
    const delta = NUDGE[key];
    if (delta && this.hasSelection && !this.dragStart && !this.selecting) {
      const from = this.selectBox!.topLeft();
      this.beginDrag(from);
      this.moveDrag(from.add(delta));
      this.end();
      return true;
    }
    return false;
  }

  /** Switching away drops the selection and any in-progress state. */
  cleanup(): void {
    this.selectedCells = [];
    this.selectBox = null;
    this.activeBox = null;
    this.attachments = [];
    this.lineReshape = null;
    this.moveTool = null;
    this.dragStart = null;
    this.dragEnd = null;
    this.selecting = false;
    this.canvas.clearSelection();
    this.canvas.clearScratch();
  }
}

File: src/core/tools/select.rs
//! Entity-aware select & move tool.
//!
//! Ported from ASCIIFlow (`client/draw/select.ts`), MIT © Lewis Hemens, plus
//! rsdia's keyboard nudge, clipboard and paste.

use indexmap::IndexMap;

use crate::core::canvas::Canvas;
use crate::core::entity::{
    cells_in_box, detect_line_tip, detect_word, find_box, move_box_with_attachments, move_cells,
    trace_box_attachments, trace_line_from_tip, BoxAttachment,
};
use crate::core::glyphs::is_box_drawing;
use crate::core::grid::{bounding_box, Bounds};
use crate::core::layer::{Layer, LayerView, StackedLayers, ERASE};
use crate::core::route::{arrow_head, connect_endpoints, line};
use crate::core::snap::snap;
use crate::core::text::{layer_to_text, text_to_layer};
use crate::core::vector::Pos;
use std::collections::HashSet;

use super::moveline::MoveTool;
use super::tool::{HoverHint, Key, Mods, Tool};

struct LineReshape {
    cells: Vec<Pos>,
    anchor: Pos,
    is_arrow: bool,
    horizontal_segment: bool,
    moved: bool,
}

fn nudge(key: Key) -> Option<Pos> {
    match key {
        Key::Up => Some(Pos::new(0, -1)),
        Key::Down => Some(Pos::new(0, 1)),
        Key::Left => Some(Pos::new(-1, 0)),
        Key::Right => Some(Pos::new(1, 0)),
        _ => None,
    }
}

fn dedupe(cells: Vec<Pos>) -> Vec<Pos> {
    let mut seen: IndexMap<Pos, ()> = IndexMap::new();
    for c in cells {
        seen.insert(c, ());
    }
    seen.into_keys().collect()
}

#[derive(Default)]
pub struct SelectTool {
    pub select_box: Option<Bounds>,
    selected_cells: Vec<Pos>,
    selecting: bool,
    select_anchor: Option<Pos>,
    drag_start: Option<Pos>,
    drag_end: Option<Pos>,
    move_tool: Option<MoveTool>,
    line_reshape: Option<LineReshape>,
    active_box: Option<Bounds>,
    attachments: Vec<BoxAttachment>,
}

impl SelectTool {
    pub fn has_selection(&self, canvas: &Canvas) -> bool {
        !self.selected_cells.is_empty() && canvas.selection.is_some()
    }

    fn in_selection(&self, canvas: &Canvas, p: Pos) -> bool {
        self.select_box.is_some_and(|b| b.contains(p)) && canvas.selection.is_some()
    }

    fn set_selection(&mut self, canvas: &mut Canvas, cells: Vec<Pos>, keep_scratch: bool) {
        self.selected_cells = dedupe(cells);
        self.select_box = bounding_box(self.selected_cells.iter().copied());
        self.active_box = None;
        if !keep_scratch {
            canvas.clear_scratch();
        }
        canvas.set_selection(self.select_box);
    }

    fn start_select(&mut self, canvas: &mut Canvas, p: Pos) {
        self.selecting = true;
        self.select_anchor = Some(p);
        self.select_box = Some(Bounds::new(p, p));
        self.selected_cells = Vec::new();
        self.active_box = None;
        canvas.set_selection(self.select_box);
    }

    fn move_select(&mut self, canvas: &mut Canvas, p: Pos) {
        let anchor = self.select_anchor.expect("a selection drag has an anchor");
        self.select_box = Some(Bounds::new(anchor, p));
        canvas.set_selection(self.select_box);
    }

    fn finish_select(&mut self, canvas: &mut Canvas) {
        let Some(b) = self.select_box else { return };
        self.selected_cells = cells_in_box(&canvas.committed, b);
        // A rubber-band selection moves like a box: lines crossing its edge reflow.
        self.active_box = self.select_box;
    }

    fn begin_drag(&mut self, canvas: &mut Canvas, p: Pos) {
        self.drag_start = Some(p);
        self.drag_end = Some(p);
        self.attachments = match self.active_box {
            Some(b) => trace_box_attachments(&canvas.committed, b),
            None => Vec::new(),
        };
    }

    fn move_drag(&mut self, canvas: &mut Canvas, p: Pos) {
        self.drag_end = Some(p);
        let Some(start) = self.drag_start else { return };
        let delta = p.subtract(start);
        if let Some(b) = self.active_box {
            canvas.set_scratch(move_box_with_attachments(
                &canvas.committed,
                b,
                &self.attachments,
                delta,
            ));
            canvas.set_selection(Some(b.translate(delta)));
        } else {
            canvas.set_scratch(move_cells(&canvas.committed, &self.selected_cells, delta));
            canvas.set_selection(bounding_box(
                self.selected_cells.iter().map(|c| c.add(delta)),
            ));
        }
    }

    fn reshape_tip(&mut self, canvas: &mut Canvas, target: Pos, m: Mods) {
        let Some(r) = self.line_reshape.as_mut() else {
            return;
        };
        r.moved = true;
        let cells = r.cells.clone();
        let anchor = r.anchor;
        let is_arrow = r.is_arrow;
        let horizontal_segment = r.horizontal_segment;

        let mut base = canvas.committed.clone();
        for c in &cells {
            base.delete(*c);
        }

        let mut layer = Layer::new();
        for c in &cells {
            layer.set(*c, ERASE);
        }

        if anchor != target {
            // Leave the pivot along the line's axis, then bend toward the cursor.
            let horizontal_first = horizontal_segment != m.flip;
            layer.set_from(&line(anchor, target, horizontal_first));
            if is_arrow {
                layer.set(target, arrow_head(anchor, target, horizontal_first));
            }
            let ends = if is_arrow {
                vec![anchor]
            } else {
                vec![anchor, target]
            };
            connect_endpoints(&mut layer, &base, &ends);
        }

        layer.set_from(&snap(&layer, &base, &HashSet::new()));
        canvas.set_scratch(layer);
        canvas.clear_selection();
    }

    /// Text of the current selection, or `None`.
    pub fn copy_selection(&self, canvas: &Canvas) -> Option<String> {
        let b = self.select_box?;
        canvas.selection.as_ref()?;
        let mut cells = Layer::new();
        for c in &self.selected_cells {
            if let Some(v) = canvas.committed.get(*c) {
                cells.set(*c, v);
            }
        }
        Some(layer_to_text(
            &StackedLayers::new(vec![&cells]),
            Some(b),
            false,
        ))
    }

    /// Erases the selected content as one undo step.
    pub fn erase_selection(&mut self, canvas: &mut Canvas) -> bool {
        if self.selected_cells.is_empty() {
            return false;
        }
        let mut layer = Layer::new();
        for c in &self.selected_cells {
            layer.set(*c, ERASE);
        }
        layer.set_from(&snap(&layer, &canvas.committed, &HashSet::new()));
        let changed = canvas.commit(layer);
        self.selected_cells = Vec::new();
        self.select_box = None;
        self.active_box = None;
        canvas.clear_selection();
        changed
    }

    /// Pastes text with its top-left at `at`, committed, and selects it.
    pub fn paste(&mut self, canvas: &mut Canvas, text: &str, at: Pos) {
        let pasted = text_to_layer(text, at);
        if pasted.is_empty() {
            return;
        }
        let mut layer = pasted.clone();
        layer.set_from(&snap(&pasted, &canvas.committed, &HashSet::new()));
        canvas.commit(layer);
        let cells: Vec<Pos> = pasted.keys();
        self.set_selection(canvas, cells, false);
    }
}

impl Tool for SelectTool {
    fn start(&mut self, canvas: &mut Canvas, p: Pos, _m: Mods) {
        let value = canvas.committed.get(p);

        if !self.selected_cells.is_empty() && self.in_selection(canvas, p) {
            self.begin_drag(canvas, p);
            return;
        }

        if let Some(tip) = detect_line_tip(&canvas.committed, p) {
            let trace = trace_line_from_tip(&canvas.committed, tip.tip, tip.body_dir);
            self.line_reshape = Some(LineReshape {
                cells: trace.cells,
                anchor: trace.anchor,
                is_arrow: tip.arrow.is_some(),
                horizontal_segment: tip.axis == crate::core::entity::Axis::Horizontal,
                moved: false,
            });
            return;
        }

        if let Some(word) = detect_word(&canvas.committed, p) {
            self.set_selection(canvas, word, false);
            self.begin_drag(canvas, p);
            return;
        }

        if value.is_some_and(is_box_drawing) {
            let mut move_tool = MoveTool::default();
            move_tool.start(canvas, p, Mods::NONE);
            self.move_tool = Some(move_tool);
            return;
        }

        if let Some(b) = find_box(&canvas.committed, p) {
            let cells = cells_in_box(&canvas.committed, b);
            self.set_selection(canvas, cells, false);
            self.active_box = Some(b);
            self.begin_drag(canvas, p);
            return;
        }

        self.start_select(canvas, p);
    }

    fn move_to(&mut self, canvas: &mut Canvas, p: Pos, m: Mods) {
        if self.line_reshape.is_some() {
            self.reshape_tip(canvas, p, m);
        } else if self.drag_start.is_some() {
            self.move_drag(canvas, p);
        } else if let Some(tool) = self.move_tool.as_mut() {
            tool.move_to(canvas, p, m);
        } else if self.selecting {
            self.move_select(canvas, p);
        }
    }

    fn end(&mut self, canvas: &mut Canvas) {
        if let Some(r) = self.line_reshape.take() {
            if r.moved {
                canvas.commit_scratch();
            } else {
                canvas.clear_scratch();
            }
        } else if let (Some(start), Some(end)) = (self.drag_start, self.drag_end) {
            let delta = end.subtract(start);
            let box_at_drag = self.active_box;
            canvas.commit_scratch();
            if let Some(b) = box_at_drag {
                let moved = b.translate(delta);
                self.selected_cells = cells_in_box(&canvas.committed, moved);
                self.select_box = Some(moved);
                self.active_box = Some(moved);
                canvas.set_selection(Some(moved));
            } else {
                let cells = self.selected_cells.iter().map(|c| c.add(delta)).collect();
                self.set_selection(canvas, cells, true);
            }
        } else if let Some(mut tool) = self.move_tool.take() {
            tool.end(canvas);
        } else if self.selecting {
            self.finish_select(canvas);
        }
        self.drag_start = None;
        self.drag_end = None;
        self.selecting = false;
    }

    fn cleanup(&mut self, canvas: &mut Canvas) {
        self.selected_cells = Vec::new();
        self.select_box = None;
        self.active_box = None;
        self.attachments = Vec::new();
        self.line_reshape = None;
        self.move_tool = None;
        self.drag_start = None;
        self.drag_end = None;
        self.selecting = false;
        canvas.clear_selection();
        canvas.clear_scratch();
    }

    fn handle_key(&mut self, canvas: &mut Canvas, key: Key, _m: Mods) -> bool {
        if key == Key::Backspace || key == Key::Delete {
            return self.erase_selection(canvas);
        }
        let Some(delta) = nudge(key) else {
            return false;
        };
        if self.has_selection(canvas) && self.drag_start.is_none() && !self.selecting {
            let from = self.select_box.expect("a selection has a box").top_left();
            self.begin_drag(canvas, from);
            self.move_drag(canvas, from.add(delta));
            self.end(canvas);
            return true;
        }
        false
    }

    fn hover_hint(&self, canvas: &Canvas, p: Pos, _m: Mods) -> HoverHint {
        let committed = &canvas.committed;
        if !self.selected_cells.is_empty() && self.in_selection(canvas, p) {
            return HoverHint::Move;
        }
        if detect_line_tip(committed, p).is_some() {
            return HoverHint::Crosshair;
        }
        if detect_word(committed, p).is_some() {
            return HoverHint::Move;
        }
        if committed.get(p).is_some_and(is_box_drawing) {
            return MoveTool::default().hover_hint(canvas, p, Mods::NONE);
        }
        if find_box(committed, p).is_some() {
            return HoverHint::Move;
        }
        HoverHint::Default
    }
}
